include/camp/device.hpp

```diff
@@ -0,0 +1,311 @@
+#ifndef __CAMP_DEVICES_HPP
+#define __CAMP_DEVICES_HPP
+
+#include <cstring>
+#include <memory>
+
+#include <cuda_runtime.h>
+namespace camp
+{
+namespace devices
+{
+
+  enum class Platform {
+    undefined = 0,
+    host = 1,
+    omp = 2,
+    tbb = 4,
+    cuda = 8,
+    // omp_target = 16, not sure this is a meaningful difference
+    hip = 32
+  };
+
+  class CudaEvent
+  {
+    public:
+      CudaEvent(cudaStream_t stream){ 
+	cudaEventCreateWithFlags(&m_event, cudaEventDisableTiming); 
+	cudaEventRecord(m_event, stream); 
+      }
+      bool check() const { return (cudaEventQuery(m_event) == cudaSuccess); }
+      void wait() const { while(!check()){} }
+    private:
+      cudaEvent_t m_event;
+  };
+
+  class HostEvent
+  {
+    public:
+      HostEvent() {}
+      bool check() const { return true; }
+      void wait() const {}
+    private:
+  };
+
+  class Event
+  {
+    public:
+      Event(){}
+      template<typename T>
+      Event(T&& value){ m_value.reset(new EventModel<T>(value));}
+
+      bool check() const { return m_value->check(); }
+      void wait() const { m_value->wait(); }
+
+      template<typename T>
+      T* get() {
+	auto result = dynamic_cast<EventModel<T>*>(m_value.get());
+	if (result ==nullptr)
+	{
+	  std::runtime_error("Incompatible Event type get cast.");
+	}
+	return result->get();
+      }
+
+    private:
+      class EventConcept {
+	public:
+	  virtual ~EventConcept(){}
+	  virtual bool check() const = 0;
+	  virtual void wait() const = 0;
+      };
+
+      template<typename T>
+      class EventModel : public EventConcept {
+	public:
+	  EventModel(T const& modelVal) : m_modelVal(modelVal) {}
+	  bool check() const override { return m_modelVal.check(); }
+	  void wait() const override { m_modelVal.wait(); }
+	  T *get() { return &m_modelVal; }
+	private:
+	  T m_modelVal;
+      };
+
+      std::unique_ptr<EventConcept> m_value;
+  };
+
+  class Cuda 
+  {
+    static cudaStream_t get_a_stream(int num)
+    {
+      // TODO consider pool size
+      static cudaStream_t streams[16] = {};
+      static int previous = 0;
+      // TODO deal with parallel init
```

Comment:
I realize I'm the one that left this in here, but it would be good to squash this TODO out before the merge.  A moving the for-loop into a lambda invoked by [std::call_once](https://en.cppreference.com/w/cpp/thread/call_once) ought to take care of it.